From 091ea27cb4989fe4692b2af095d4d99b2c596ce8 Mon Sep 17 00:00:00 2001
From: Michael Allman <michael@videoamp.com>
Date: Mon, 5 Dec 2016 15:27:14 -0800
Subject: [PATCH] PARQUET-783: Close the underlying stream when an
 H2SeekableInputStream is closed

This PR addresses https://issues.apache.org/jira/browse/PARQUET-783.

`ParquetFileReader` opens a `SeekableInputStream` to read a footer. In the process, it opens a new `FSDataInputStream` and wraps it. However, `H2SeekableInputStream` does not override the `close` method. Therefore, when `ParquetFileReader` closes it, the underlying `FSDataInputStream` is not closed. As a result, these stale connections can exhaust a clusters' data nodes' connection resources and lead to mysterious HDFS read failures in HDFS clients, e.g.

```
org.apache.hadoop.hdfs.BlockMissingException: Could not obtain block: BP-905337612-172.16.70.103-1444328960665:blk_1720536852_646811517
```

Author: Michael Allman <michael@videoamp.com>

Closes #388 from mallman/parquet-783-close_underlying_inputstream and squashes the following commits:

f4b27c1 [Michael Allman] PARQUET-783 Close the underlying stream when an H2SeekableInputStream is closed
---
 .../apache/parquet/hadoop/util/H2SeekableInputStream.java    | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/H2SeekableInputStream.java b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/H2SeekableInputStream.java
index a7065465f9..ec4567e4e4 100644
--- a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/H2SeekableInputStream.java
+++ b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/H2SeekableInputStream.java
@@ -44,6 +44,11 @@ public H2SeekableInputStream(FSDataInputStream stream) {
     this.reader = new H2Reader();
   }
 
+  @Override
+  public void close() throws IOException {
+    stream.close();
+  }
+
   @Override
   public long getPos() throws IOException {
     return stream.getPos();
